### PR TITLE
CFE-2918: Fixed various memory leaks

### PR DESCRIPTION
**Pull request by olehermanse**

### CODE REVIEW

**Comment by olehermanse** [comment]
@cf-bottom jenkins

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/1338/)

https://ci.cfengine.com/job/pr-pipeline/1338/

**Comment by olehermanse** [comment]
@cf-bottom Let's do jenkins once more

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/1340/)

https://ci.cfengine.com/job/pr-pipeline/1340/

File: cf-key/cf-key-functions.c [hunk 1]
```diff
@@ -271,6 +229,8 @@ bool KeepKeyPromises(const char *public_key_file, const char *private_key_file,
 
     Log(LOG_LEVEL_VERBOSE, "Writing private key to '%s'", private_key_file);
 
+    const EVP_CIPHER *cipher = EVP_des_ede3_cbc();
+
     if (!PEM_write_RSAPrivateKey(fp, pair, cipher, (void *)PRIVKEY_PASSPHRASE,
                                  PRIVKEY_PASSPHRASE_LEN, NULL, NULL))
     {
```
Comment (vpodzime) [inline cf-key/cf-key-functions.c]: The block below this needs to do `fclose(fp)` before return.

File: cf-key/cf-key-functions.c [hunk 2]
```diff
@@ -209,60 +209,18 @@ int RemoveKeys(const char *input, bool must_be_coherent)
...
         Log(LOG_LEVEL_ERR, "Couldn't open private key file '%s' (open: %s)", private_key_file, GetErrorStr());
         return false;
     }
 
-    if ((fp = fdopen(fd, "w")) == NULL)
+    FILE *fp = fdopen(fd, "w");
+    if (fp == NULL)
     {
         Log(LOG_LEVEL_ERR, "Error while writing private key file '%s' (fdopen: %s)", private_key_file, GetErrorStr());
```
[the remark this answers is not in the document]
Comment (vpodzime) [reply cf-key/cf-key-functions.c]: Same below for the other file.

File: cf-key/cf-key-functions.c [hunk 3]
```diff
@@ -331,6 +292,60 @@ bool KeepKeyPromises(const char *public_key_file, const char *private_key_file,
...
+    RSA *pair = RSA_new();
+    BIGNUM *rsa_bignum = BN_new();
+    if (pair != NULL && rsa_bignum != NULL)
+    {
+        BN_set_word(rsa_bignum, RSA_F4);
+        int res = RSA_generate_key_ex(pair, key_size, rsa_bignum, NULL);
+        if (res == 0)
+        {
+            RSA_free(pair);
+            pair = NULL;
```
Comment (vpodzime) [inline cf-key/cf-key-functions.c]: You can use your new macros here.

File: cf-key/cf-key-functions.c [hunk 4]
```diff
@@ -331,6 +292,60 @@ bool KeepKeyPromises(const char *public_key_file, const char *private_key_file,
...
+        }
+    }
+    else
+    {
+        RSA_free(pair);
```
Comment (vpodzime) [inline cf-key/cf-key-functions.c]: I take it this is safe to call with `NULL`.?

Comment (olehermanse) [reply cf-key/cf-key-functions.c]: Yes.

File: cf-key/cf-key-functions.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (olehermanse) [reply cf-key/cf-key-functions.c]: I didn't introduce it ;) Also, I don't think it's bad for a user level error message. Error while opening would be more confusing.

**Comment by olehermanse** [comment]
@cf-bottom jenkins

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/1341/)

https://ci.cfengine.com/job/pr-pipeline/1341/

**Comment by olehermanse** [comment]
@vpodzime thanks for your comments, I have fixed them ;)

**Pull request merged**